raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -350,20 +325,10 @@ private void resetConnections() {
     }
 
     private void onBecomeLeader(LeaderState state) {
-        stateMachine.becomeLeader(quorum.epoch());
-        updateLeaderEndOffsetAndTimestamp(state);
-
         // Add a control message for faster high watermark advance.
         final long currentTimeMs = time.milliseconds();
-        appendControlRecord(MemoryRecords.withLeaderChangeMessage(
-            currentTimeMs,
-            quorum.epoch(),
-            new LeaderChangeMessage()
-                .setLeaderId(state.election().leaderId())
-                .setVoters(
-                    state.followers().stream().map(
-                        follower -> new Voter().setVoterId(follower)).collect(Collectors.toList())))
-        );
+        appendLeaderChangeMessage(state, currentTimeMs);
+        updateLeaderEndOffsetAndTimestamp(state);
```

Comment:
The ordering of this prior to the append of the leader change message was a bug which caused a few test cases to be off.